src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,392 @@
+      request.headers.forEach((value, key) => {
+        headers[key] = value;
+      });
+
+      const serializedRequest: SerializedRequest = {
+        url: requestClone.url,
+        method: requestClone.method,
+        headers: JSON.stringify(headers),
+        body: serializedBody.body,
+        bodyType: serializedBody.type,
+        timestamp: Date.now(),
+        type,
+        stepKey
+      };
+
+      const { tx, store } = await this.getDbTransaction('readwrite');
+      const requestPromise = new Promise<void>((resolve, reject) => {
+        tx.oncomplete = () => resolve();
+        tx.onerror = () => reject(tx.error);
+      });
+
+      store.add(serializedRequest);
+      await requestPromise;
+    } catch (error) {
+      console.error('Error adding request to IndexedDB:', error);
+    }
+  }
+
+  // Helper function to serialize request body based on its type
+  private async serializeRequestBody(
+    body: any
+  ): Promise<SerializedRequestBody> {
+    if (body === null) {
+      return { type: 'text', body: '' };
+    }
+    if (body instanceof Blob) {
+      return { type: 'blob', body: await body.arrayBuffer() }; // Convert Blob to ArrayBuffer for storage
+    } else if (body instanceof FormData) {
+      // FormData needs to be converted to a plain object
+      const formDataObj: Record<string, any> = {};
+      body.forEach((value, key) => {
+        formDataObj[key] = value;
+      });
+      return { type: 'formData', body: formDataObj };
+    } else if (body instanceof ArrayBuffer) {
+      return { type: 'arrayBuffer', body: body };
+    } else {
+      // Assume it's text or JSON
+      return { type: 'text', body: await body.text() };
+    }
+  }
+
+  // Replay requests from IndexedDB
+  public async replayRequests() {
+    if (!this.indexedDBSupported) {
+      return;
+    }
+
+    if (!navigator.onLine) {
+      // If offline, block until requests are replayed from online event handler
+      await new Promise((resolve) => this._addOnlineSignal(resolve));
+      return;
+    }
+
+    if (this.isReplayingRequests) {
+      return;
+    }
+
+    this.isReplayingRequests = true;
+
+    try {
+      const { store } = await this.getDbTransaction('readwrite');
+
+      const allRequests: SerializedRequest[] = await new Promise<
+        SerializedRequest[]
+      >((resolve) => {
+        const requests: SerializedRequest[] = [];
+        store.openCursor().onsuccess = (event: Event) => {
+          const cursor = (event.target as IDBRequest<IDBCursorWithValue | null>)
+            .result;
+          if (cursor) {
+            requests.push(cursor.value as SerializedRequest);
+            cursor.continue();
+          } else {
+            resolve(requests);
+          }
+        };
+      });
+
+      const submitCustomRequests = allRequests.filter(
+        (req) => req.type === 'submitCustom'
+      );
+      const submitStepRequests = allRequests.filter(
+        (req) => req.type === 'submitStep'
```

Comment:
you could just have this and the submitCustom requests all have the `submit` type, then filter by req.type === 'submit'